// @nicotordev/flowcl-pagos/dist/clients/flow.merchants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const flow_utils_1 = require("../utils/flow.utils");
const errors_1 = require("../errors");
const qs_1 = __importDefault(require("qs"));
/**
 * Permite gestionar los comercios asociados
 *
 */
class FlowMerchants {
    /**
     * Constructor de la clase FlowClient.
     * @param {string} apiKey Clave de API proporcionada por Flow.
     * @param {string} secretKey Clave secreta proporcionada por Flow.
     * @param {string} baseURL URL base de la API de Flow.
     * @throws {FlowAuthenticationError} Si no se proporciona apiKey o secretKey.
     */
    constructor(apiKey, secretKey, baseURL) {
        this.createAssociatedMerchant = this._createAssociatedMerchant.bind(this);
        /**
         * Este método permite modificar un comercio asociado previamente creado en Flow
         * @param {FlowEditAssociatedMerchantRequest} data Datos del comercio asociado a modificar
         * @returns {Promise<FlowEditAssociatedMerchantResponse>} Respuesta de la API de Flow
         * @throws {FlowAPIError} Si hay problemas con la API de Flow.
         * @throws {FlowCreateAssociatedMerchantError} Si hay problemas al modificar el comercio asociado.
         */
        this.editAssociatedMerchant = this._editAssociatedMerchant.bind(this);
        /**
         * Este método permite eliminar un comercio asociado previamente creado en Flow
         * @param {string} id ID del comercio asociado a eliminar
         * @returns {Promise<FlowDeleteAssociatedMerchantResponse>} Respuesta de la API de Flow
         * @throws {FlowAPIError} Si hay problemas con la API de Flow.
         * @throws {FlowCreateAssociatedMerchantError} Si hay problemas al eliminar el comercio asociado.
         */
        this.deleteAssociatedMerchant = this._deleteAssociatedMerchant.bind(this);
        /**
         * Este método permite obtener la información de un comercio asociado previamente creado en Flow
         *
         */
        this.getAssociatedMerchant = this._getAssociatedMerchant.bind(this);
        /**
         * Permite obtener la lista de comercios paginada de acuerdo a los parámetros de paginación. Además, se puede definir los siguientes filtros:
         * filter: filtro por nombre del comercio asociad
         * status: filtro por estado del comercio asociado
         * @param {FlowGetAssociatedMerchantsRequest} data Datos de la petición.
         * @returns {Promise<FlowGetAssociatedMerchantsResponse>} Respuesta de la API.
         * @throws {FlowAPIError} Si hay problemas con la API de Flow.
         * @throws {FlowDeleteAssociatedMerchantError} Si hay problemas al realizar la petición.
         */
        this.getAssociatedMerchants = this._getAssociatedMerchants.bind(this);
        if (!apiKey || !secretKey) {
            throw new errors_1.FlowAuthenticationError();
        }
        this.apiKey = apiKey;
        this.secretKey = secretKey;
        // Crear una instancia de Axios con la configuración base
        this.axiosInstance = axios_1.default.create({
            baseURL: `${baseURL}/merchant`,
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded',
            },
        });
    }
    /**
     * Realiza una petición a la API de Flow.
     * @param {string} endpoint URL del endpoint de la API.
     * @param {Record<string, unknown>} data Datos a enviar en la petición.
     * @param {'post' | 'get'} method Método de la petición (POST o GET).
     * @param {(err: unknown) => never} error Error a lanzar en caso de error.
     * @returns {Promise<T>} Respuesta de la API.
     * @throws {FlowAPIError} Si hay problemas con la API de Flow.
     * @throws {Error} Si hay problemas al realizar la petición.
     */
    async request(endpoint, data, method = 'post', error) {
        try {
            const allData = {
                ...data,
                apiKey: this.apiKey,
            };
            const formData = (0, flow_utils_1.generateFormData)(allData, this.secretKey);
            const formDataSearchParams = new URLSearchParams(formData);
            const response = method === 'post'
                ? await this.axiosInstance.post(`${endpoint}`, qs_1.default.stringify(formData))
                : await this.axiosInstance.get(`${endpoint}?${formDataSearchParams}`);
            return response.data;
        }
        catch (err) {
            if (axios_1.default.isAxiosError(err)) {
                console.error(err.response?.data);
                throw new errors_1.FlowAPIError(err.response?.status || 500, err.message);
            }
            error(err);
        }
    }
    /**
     * Este método permite crear un nuevo comercio asociado en Flow
     * @param {FlowCreateAssociatedMerchantRequest} data Datos del comercio asociado a crear
     * @returns {Promise<FlowCreateAssociatedMerchantRequest>} Respuesta de la API de Flow
     * @throws {FlowAPIError} Si hay problemas con la API de Flow.
     * @throws {FlowCreateAssociatedMerchantError} Si hay problemas al crear el comercio asociado.
     */
    async _createAssociatedMerchant(data) {
        return await this.request('/create', data, 'post', (er) => {
            throw new errors_1.FlowCreateAssociatedMerchantError(er.message);
        });
    }
    /**
     * Este método permite modificar un comercio asociado previamente creado en Flow
     * @param {FlowEditAssociatedMerchantRequest} data Datos del comercio asociado a modificar
     * @returns {Promise<FlowEditAssociatedMerchantResponse>} Respuesta de la API de Flow
     * @throws {FlowAPIError} Si hay problemas con la API de Flow.
     * @throws {FlowCreateAssociatedMerchantError} Si hay problemas al modificar el comercio asociado.
     */
    async _editAssociatedMerchant(data) {
        return await this.request('/edit', data, 'post', (er) => {
            throw new errors_1.FlowCreateAssociatedMerchantError(er.message);
        });
    }
    /**
     * Este método permite eliminar un comercio asociado previamente creado en Flow
     * @param {string} id ID del comercio asociado a eliminar
     * @returns {Promise<FlowDeleteAssociatedMerchantResponse>} Respuesta de la API de Flow
     * @throws {FlowAPIError} Si hay problemas con la API de Flow.
     * @throws {FlowCreateAssociatedMerchantError} Si hay problemas al eliminar el comercio asociado.
     */
    async _deleteAssociatedMerchant(id) {
        return await this.request('/delete', { id }, 'post', (er) => {
            throw new errors_1.FlowDeleteAssociatedMerchantError(er.message);
        });
    }
    /**
     * Este método permite obtener la información de un comercio asociado previamente creado en Flow
     *
     */
    async _getAssociatedMerchant(id) {
        return await this.request('/get', { id }, 'get', (er) => {
            throw new errors_1.FlowDeleteAssociatedMerchantError(er.message);
        });
    }
    /**
     * Permite obtener la lista de comercios paginada de acuerdo a los parámetros de paginación. Además, se puede definir los siguientes filtros:
     * filter: filtro por nombre del comercio asociad
     * status: filtro por estado del comercio asociado
     * @param {FlowGetAssociatedMerchantsRequest} data Datos de la petición.
     * @returns {Promise<FlowGetAssociatedMerchantsResponse>} Respuesta de la API.
     * @throws {FlowAPIError} Si hay problemas con la API de Flow.
     * @throws {FlowDeleteAssociatedMerchantError} Si hay problemas al realizar la petición.
     */
    async _getAssociatedMerchants(data) {
        return await this.request('/list', data, 'get', (er) => {
            throw new errors_1.FlowDeleteAssociatedMerchantError(er.message);
        });
    }
}
exports.default = FlowMerchants;
