// @nicotordev/flowcl-pagos/dist/clients/flow.settlement.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const errors_1 = require("../errors");
const flow_utils_1 = require("../utils/flow.utils");
const qs_1 = __importDefault(require("qs"));
/**
 * Permite obtener las liquidaciones de pagos efectuadas por Flow
 */
class FlowSettlements {
    /**
     * Constructor de la clase FlowClient.
     * @param {string} apiKey Clave de API proporcionada por Flow.
     * @param {string} secretKey Clave secreta proporcionada por Flow.
     * @param {string} baseURL URL base de la API de Flow.
     * @throws {FlowAuthenticationError} Si no se proporciona apiKey o secretKey.
     */
    constructor(apiKey, secretKey, baseURL) {
        /**
         * Este método se utiliza para obtener el(los) encabezado(s) de liquidación(es) dentro del rango de fechas ingresado (permite filtrar también por la moneda). Para obtener la liquidación completa (encabezado y detalles) utilizar el servicio /settlement/getByIdv2
         * @param {FlowGetSettlementHeadersByDateRangeRequest} data Datos de la petición.
         * @returns {Promise<FlowGetSettlementHeadersByDateRangeResponse>} Respuesta de la API.
         * @throws {FlowAPIError} Si hay problemas con la API de Flow.
         * @throws {FlowGetLiquidationsByDateRangeError} Si hay problemas al realizar la petición.
         */
        this.getLiquidationsByDateRange = this._getLiquidationsByDateRange.bind(this);
        /**
         * Este método se utiliza para obtener el objeto Settlement correspondiente al identificador
         * @param {string} id Identificador de la liquidación.
         * @returns {Promise<FlowGetLiquidationByIdResponse>} Respuesta de la API.
         * @throws {FlowAPIError} Si hay problemas con la API de Flow.
         * @throws {FlowGetLiquidationsByDateRangeError} Si hay problemas al realizar la petición.
         */
        this.getLiquidationById = this._getLiquidationById.bind(this);
        if (!apiKey || !secretKey) {
            throw new errors_1.FlowAuthenticationError();
        }
        this.apiKey = apiKey;
        this.secretKey = secretKey;
        // Crear una instancia de Axios con la configuración base
        this.axiosInstance = axios_1.default.create({
            baseURL: `${baseURL}/settlement`,
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded',
            },
        });
    }
    /**
     * Realiza una petición a la API de Flow.
     * @param {string} endpoint URL del endpoint de la API.
     * @param {Record<string, unknown>} data Datos a enviar en la petición.
     * @param {'post' | 'get'} method Método de la petición (POST o GET).
     * @param {(err: unknown) => never} error Error a lanzar en caso de error.
     * @returns {Promise<T>} Respuesta de la API.
     * @throws {FlowAPIError} Si hay problemas con la API de Flow.
     * @throws {Error} Si hay problemas al realizar la petición.
     */
    async request(endpoint, data, method = 'post', error) {
        try {
            const allData = {
                ...data,
                apiKey: this.apiKey,
            };
            const formData = (0, flow_utils_1.generateFormData)(allData, this.secretKey);
            const formDataSearchParams = new URLSearchParams(formData);
            const response = method === 'post'
                ? await this.axiosInstance.post(`${endpoint}`, qs_1.default.stringify(formData))
                : await this.axiosInstance.get(`${endpoint}?${formDataSearchParams}`);
            return response.data;
        }
        catch (err) {
            if (axios_1.default.isAxiosError(err)) {
                throw new errors_1.FlowAPIError(err.response?.status || 500, err.message);
            }
            error(err);
        }
    }
    /**
     * Este método se utiliza para obtener el(los) encabezado(s) de liquidación(es) dentro del rango de fechas ingresado (permite filtrar también por la moneda). Para obtener la liquidación completa (encabezado y detalles) utilizar el servicio /settlement/getByIdv2
     * @param {FlowGetSettlementHeadersByDateRangeRequest} data Datos de la petición.
     * @returns {Promise<FlowGetSettlementHeadersByDateRangeResponse>} Respuesta de la API.
     * @throws {FlowAPIError} Si hay problemas con la API de Flow.
     * @throws {FlowGetLiquidationsByDateRangeError} Si hay problemas al realizar la petición.
     */
    async _getLiquidationsByDateRange(data) {
        return await this.request('/search', data, 'post', (err) => {
            throw new errors_1.FlowGetLiquidationsByDateRangeError(err.message);
        });
    }
    /**
     * Este método se utiliza para obtener el objeto Settlement correspondiente al identificador
     * @param {string} id Identificador de la liquidación.
     * @returns {Promise<FlowGetLiquidationByIdResponse>} Respuesta de la API.
     * @throws {FlowAPIError} Si hay problemas con la API de Flow.
     * @throws {FlowGetLiquidationsByDateRangeError} Si hay problemas al realizar la petición.
     */
    async _getLiquidationById(id) {
        return await this.request(`/getByIdv2`, { id }, 'get', (err) => {
            throw new errors_1.FlowGetLiquidationsByDateRangeError(err.message);
        });
    }
}
exports.default = FlowSettlements;
