// @nicotordev/flowcl-pagos/dist/clients/flow.coupons.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("../errors");
const axios_1 = __importDefault(require("axios"));
const flow_utils_1 = require("../utils/flow.utils");
const qs_1 = __importDefault(require("qs"));
/**
 * Permite crear cupones de descuento para ser aplicados a suscripciones o clientes
 */
class FlowCoupons {
    /**
     * Constructor de la clase FlowClient.
     * @param apiKey Clave de API proporcionada por Flow.
     * @param secretKey Clave secreta proporcionada por Flow.
     * @param baseURL URL base de la API de Flow.
     * @throws FlowAuthenticationError Si no se proporciona apiKey o secretKey.
     */
    constructor(apiKey, secretKey, baseURL) {
        /**
         * Este servicio permite crear un cupón de descuento
         * @param {FlowCreateDiscountCouponRequest} data Datos para crear el cupón de descuento
         * @returns {Promise<FlowCreateDiscountCouponResponse>} Respuesta de la API
         * @throws {FlowCreateDiscountCouponError} Si hay problemas al crear el cupón de descuento
         * @throws {FlowAPIError} Si hay problemas con la API de Flow
         */
        this.create = this.createDiscountCoupon.bind(this);
        /**
         * Este servicio permite editar un cupón de descuento. Sólo se puede editar el nombre de un cupón.
         * @param {FlowEditDiscountCouponRequest} data Datos para editar el cupón de descuento
         * @returns {Promise<FlowEditDiscountCouponResponse>} Respuesta de la API
         * @throws {FlowEditDiscountCouponError} Si hay problemas al editar el cupón de descuento
         * @throws {FlowAPIError} Si hay problemas con la API de Flow
         */
        this.edit = this.editDiscountCoupon.bind(this);
        /**
         * Este servicio permite eliminar un cupón de descuento. Eliminar un cupón de descuento no elimina los descuentos aplicados a clientes o suscripciones, sólo no permite volver a aplicar este cupón
         * @param {string} couponId ID del cupón de descuento
         * @returns {Promise<FlowDeleteDiscountCouponResponse>} Respuesta de la API
         * @throws {FlowEditDiscountCouponError} Si hay problemas al eliminar el cupón de descuento
         * @throws {FlowAPIError} Si hay problemas con la API de Flow
         */
        this.delete = this.deleteDiscountCoupon.bind(this);
        /**
         * Este servicio permite obtener los datos de un cupón de descuento
         * @param {string} couponId ID del cupón de descuento
         * @returns {Promise<FlowGetDiscountCouponResponse>} Respuesta de la API
         * @throws {FlowGetDiscountCouponError} Si hay problemas al obtener el cupón de descuento
         * @throws {FlowAPIError} Si hay problemas con la API de Flow
         */
        this.get = this.getDiscountCoupon.bind(this);
        /**
         * Este servicio permite la lista de cupones de descuento
         * @param {FlowListDiscountCouponsRequest} data Datos para listar los cupones de descuento
         * @returns {Promise<FlowListDiscountCouponsResponse>} Respuesta de la API
         * @throws {FlowListDiscountCouponsError} Si hay problemas al listar los cupones de descuento
         * @throws {FlowAPIError} Si hay problemas con la API de Flow
         */
        this.list = this.listDiscountCoupons.bind(this);
        if (!apiKey || !secretKey) {
            throw new errors_1.FlowAuthenticationError();
        }
        this.apiKey = apiKey;
        this.secretKey = secretKey;
        // Crear una instancia de Axios con la configuración base
        this.axiosInstance = axios_1.default.create({
            baseURL: `${baseURL}/coupon`,
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded',
            },
        });
    }
    /**
     * Realiza una petición a la API de Flow.
     * @param {string} endpoint URL del endpoint de la API.
     * @param {string} data Datos a enviar en la petición.
     * @param {'post' | 'get'} method Método de la petición (POST o GET).
     * @param {(e: unknown) => never} error Error a lanzar en caso de error.
     * @returns {Promise<T>} Respuesta de la API.
     * @throws {FlowAPIError} Si hay problemas con la API de Flow.
     * @throws {Error} Si hay problemas al realizar la petición.
     */
    async request(endpoint, data, method = 'post', error) {
        try {
            const allData = {
                ...data,
                apiKey: this.apiKey,
            };
            const formData = (0, flow_utils_1.generateFormData)(allData, this.secretKey);
            const formDataSearchParams = new URLSearchParams(formData);
            const response = method === 'post'
                ? await this.axiosInstance.post(`${endpoint}`, qs_1.default.stringify(formData))
                : await this.axiosInstance.get(`${endpoint}?${formDataSearchParams}`);
            return response.data;
        }
        catch (err) {
            if (axios_1.default.isAxiosError(err)) {
                console.error(JSON.stringify(err.response?.data, null, 2));
                throw new errors_1.FlowAPIError(err.response?.status || 500, err.message);
            }
            error(err);
        }
    }
    /**
     * Este servicio permite crear un cupón de descuento
     * @param {FlowCreateDiscountCouponRequest} data Datos para crear el cupón de descuento
     * @returns {Promise<FlowCreateDiscountCouponResponse>} Respuesta de la API
     * @throws {FlowCreateDiscountCouponError} Si hay problemas al crear el cupón de descuento
     * @throws {FlowAPIError} Si hay problemas con la API de Flow
     */
    async createDiscountCoupon(data) {
        return await this.request('/create', data, 'post', (e) => {
            throw new errors_1.FlowCreateDiscountCouponError(e.message);
        });
    }
    /**
     * Este servicio permite editar un cupón de descuento. Sólo se puede editar el nombre de un cupón.
     * @param {FlowEditDiscountCouponRequest} data Datos para editar el cupón de descuento
     * @returns {Promise<FlowEditDiscountCouponResponse>} Respuesta de la API
     * @throws {FlowEditDiscountCouponError} Si hay problemas al editar el cupón de descuento
     * @throws {FlowAPIError} Si hay problemas con la API de Flow
     */
    async editDiscountCoupon(data) {
        return await this.request('/edit', data, 'post', (e) => {
            throw new errors_1.FlowEditDiscountCouponError(e.message);
        });
    }
    /**
     * Este servicio permite eliminar un cupón de descuento. Eliminar un cupón de descuento no elimina los descuentos aplicados a clientes o suscripciones, sólo no permite volver a aplicar este cupón
     * @param {string} couponId ID del cupón de descuento
     * @returns {Promise<FlowDeleteDiscountCouponResponse>} Respuesta de la API
     * @throws {FlowEditDiscountCouponError} Si hay problemas al eliminar el cupón de descuento
     * @throws {FlowAPIError} Si hay problemas con la API de Flow
     */
    async deleteDiscountCoupon(couponId) {
        return await this.request('/delete', { couponId }, 'post', (e) => {
            throw new errors_1.FlowDeleteDiscountCouponError(e.message);
        });
    }
    /**
     * Este servicio permite obtener los datos de un cupón de descuento
     * @param {string} couponId ID del cupón de descuento
     * @returns {Promise<FlowGetDiscountCouponResponse>} Respuesta de la API
     * @throws {FlowGetDiscountCouponError} Si hay problemas al obtener el cupón de descuento
     * @throws {FlowAPIError} Si hay problemas con la API de Flow
     */
    async getDiscountCoupon(couponId) {
        return await this.request('/get', { couponId }, 'get', (e) => {
            throw new errors_1.FlowGetDiscountCouponError(e.message);
        });
    }
    /**
     * Este servicio permite la lista de cupones de descuento
     * @param {FlowListDiscountCouponsRequest} data Datos para listar los cupones de descuento
     * @returns {Promise<FlowListDiscountCouponsResponse>} Respuesta de la API
     * @throws {FlowListDiscountCouponsError} Si hay problemas al listar los cupones de descuento
     * @throws {FlowAPIError} Si hay problemas con la API de Flow
     */
    async listDiscountCoupons(data) {
        return await this.request('/list', data, 'get', (e) => {
            throw new errors_1.FlowListDiscountCouponsError(e.message);
        });
    }
}
exports.default = FlowCoupons;
