// @nicotordev/flowcl-pagos/dist/errors.js
"use strict";
// src/errors.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowListAdditionalSubscriptionItemError = exports.FlowDeleteAdditionalSubscriptionItemError = exports.FlowEditAdditionalSubscriptionItemError = exports.FlowGetAdditionalSubscriptionItemError = exports.FlowCreateAdditionalSubscriptionItemError = exports.FlowCancelScheduledPlanChangeError = exports.FlowPreviewSubscriptionPlanChangeError = exports.FlowChangeAssociatedPlanToSubscriptionError = exports.FlowRemoveItemFromSubscriptionError = exports.FlowAddItemToSubscriptionError = exports.FlowRemoveDiscountFromSubscriptionError = exports.FlowAddDiscountToSubscriptionError = exports.FlowCancelSubscriptionError = exports.FlowUpdateSubscriptionTrialDaysError = exports.FlowGetPlanSubscriptionsError = exports.FlowGetSubscriptionBySubscriptionIdError = exports.FlowCreateSubscriptionToPlanError = exports.FlowListPlansError = exports.FlowEditPlanError = exports.FlowGetPlanError = exports.FlowCreatePlanError = exports.FlowListPaginatedSubscriptionsError = exports.FlowListFailedChargesCardError = exports.FlowListChargesCardError = exports.FlowReverseChargeCardError = exports.FlowMassiveChargeCardStatusError = exports.FlowSendMassiveChargeCardError = exports.FlowSendChargeCardError = exports.FlowChargeCardError = exports.FlowDeleteCardError = exports.FlowRegisterCardStatusError = exports.FlowRegisterCardError = exports.FlowGetCustomerListError = exports.FlowGetCustomerError = exports.FlowDeleteCustomerError = exports.FlowEditCustomerError = exports.FlowCreateCustomerError = exports.FlowRefundStatusError = exports.FlowCancelRefundError = exports.FlowCreateRefundError = exports.FlowCreatePaymentByEmailError = exports.FlowStatusExtendedError = exports.FlowCreatePaymentError = exports.FlowTransactionsReceivedByDateError = exports.FlowPaymentsReceivedByDateError = exports.FlowAPIError = exports.FlowPaymentStatusError = exports.FlowOrderCreationError = exports.FlowAuthenticationError = exports.FlowError = void 0;
exports.FlowGetAssociatedMerchantError = exports.FlowDeleteAssociatedMerchantError = exports.FlowCreateAssociatedMerchantError = exports.FlowGetLiquidationsByDateRangeError = exports.FlowRetryOverdueInvoicePaymentError = exports.FlowGetOverdueInvoicesError = exports.FlowRecordExternalPaymentAndMarkInvoicePaidError = exports.FlowCancelInvoicePendingPaymentError = exports.FlowGetInvoiceDataError = exports.FlowListDiscountCouponsError = exports.FlowGetDiscountCouponError = exports.FlowDeleteDiscountCouponError = exports.FlowEditDiscountCouponError = exports.FlowCreateDiscountCouponError = void 0;
/**
 * Error base para todos los errores personalizados de FlowClient.
 */
class FlowError extends Error {
    constructor(message) {
        super(message);
        this.name = 'FlowError';
    }
}
exports.FlowError = FlowError;
/**
 * Error cuando las credenciales de API son inválidas o faltan.
 */
class FlowAuthenticationError extends FlowError {
    constructor() {
        super('API Key y Secret Key son requeridos.');
        this.name = 'FlowAuthenticationError';
    }
}
exports.FlowAuthenticationError = FlowAuthenticationError;
/**
 * Error cuando una orden de pago no se puede crear.
 */
class FlowOrderCreationError extends FlowError {
    constructor(message) {
        super(`Error al crear la orden de pago: ${message}`);
        this.name = 'FlowOrderCreationError';
    }
}
exports.FlowOrderCreationError = FlowOrderCreationError;
/**
 * Error cuando hay problemas al obtener el estado de un pago.
 */
class FlowPaymentStatusError extends FlowError {
    constructor(message) {
        super(`Error al obtener el estado del pago: ${message}`);
        this.name = 'FlowPaymentStatusError';
    }
}
exports.FlowPaymentStatusError = FlowPaymentStatusError;
/**
 * Error genérico para errores HTTP.
 */
class FlowAPIError extends FlowError {
    constructor(statusCode, message) {
        super(`Error de API (${statusCode}): ${message}`);
        this.name = 'FlowAPIError';
        this.statusCode = statusCode;
    }
}
exports.FlowAPIError = FlowAPIError;
/**
 * Error cuando hay problemas al obtener los pagos recibidos por fecha.
 */
class FlowPaymentsReceivedByDateError extends FlowError {
    constructor(message) {
        super(`Error al obtener los pagos recibidos por fecha: ${message}`);
        this.name = 'PaymentsReceivedByDateError';
    }
}
exports.FlowPaymentsReceivedByDateError = FlowPaymentsReceivedByDateError;
/**
 * Error cuando hay problemas al obtener las transacciones recibidas por fecha.
 */
class FlowTransactionsReceivedByDateError extends FlowError {
    constructor(message) {
        super(`Error al obtener las transacciones recibidas por fecha: ${message}`);
        this.name = 'TransactionsReceivedByDateError';
    }
}
exports.FlowTransactionsReceivedByDateError = FlowTransactionsReceivedByDateError;
/**
 * Error cuando hay problemas al crear un pago.
 */
class FlowCreatePaymentError extends FlowError {
    constructor(message) {
        super(`Error al crear el pago: ${message}`);
        this.name = 'FlowCreatePaymentError';
    }
}
exports.FlowCreatePaymentError = FlowCreatePaymentError;
/**
 * Error cuando hay problemas al obtener el estado extendido de un pago.
 */
class FlowStatusExtendedError extends FlowError {
    constructor(message) {
        super(`Error al obtener el estado extendido del pago: ${message}`);
        this.name = 'FlowStatusExtendedError';
    }
}
exports.FlowStatusExtendedError = FlowStatusExtendedError;
/**
 * Error cuando hay problemas al obtener el estado de un pago por email.
 */
class FlowCreatePaymentByEmailError extends FlowError {
    constructor(message) {
        super(`Error al crear el pago por email: ${message}`);
        this.name = 'FlowCreatePaymentByEmailError';
    }
}
exports.FlowCreatePaymentByEmailError = FlowCreatePaymentByEmailError;
/**
 * Error cuando hay problemas al obtener el estado de un reembolso.
 */
class FlowCreateRefundError extends FlowError {
    constructor(message) {
        super(`Error al crear el reembolso: ${message}`);
        this.name = 'FlowCreateRefundError';
    }
}
exports.FlowCreateRefundError = FlowCreateRefundError;
/**
 * Error cuando hay problemas al cancelar un reembolso.
 */
class FlowCancelRefundError extends FlowError {
    constructor(message) {
        super(`Error al cancelar el reembolso: ${message}`);
        this.name = 'FlowCancelRefundError';
    }
}
exports.FlowCancelRefundError = FlowCancelRefundError;
/**
 * Error cuando hay problemas al obtener el estado de un reembolso.
 */
class FlowRefundStatusError extends FlowError {
    constructor(message) {
        super(`Error al obtener el estado del reembolso: ${message}`);
        this.name = 'FlowRefundStatusError';
    }
}
exports.FlowRefundStatusError = FlowRefundStatusError;
/**
 * Error cuando hay problemas al obtener el estado de un reembolso por email.
 */
class FlowCreateCustomerError extends FlowError {
    constructor(message) {
        super(`Error al crear el cliente: ${message}`);
        this.name = 'FlowCreateCustomerError';
    }
}
exports.FlowCreateCustomerError = FlowCreateCustomerError;
/**
 * Error cuando hay problemas al editar un cliente.
 */
class FlowEditCustomerError extends FlowError {
    constructor(message) {
        super(`Error al editar el cliente: ${message}`);
        this.name = 'FlowEditCustomerError';
    }
}
exports.FlowEditCustomerError = FlowEditCustomerError;
/**
 * Error cuando hay problemas al eliminar un cliente.
 */
class FlowDeleteCustomerError extends FlowError {
    constructor(message) {
        super(`Error al eliminar el cliente: ${message}`);
        this.name = 'FlowDeleteCustomerError';
    }
}
exports.FlowDeleteCustomerError = FlowDeleteCustomerError;
/**
 * Error cuando hay problemas al obtener un cliente.
 */
class FlowGetCustomerError extends FlowError {
    constructor(message) {
        super(`Error al obtener el cliente: ${message}`);
        this.name = 'FlowGetCustomerError';
    }
}
exports.FlowGetCustomerError = FlowGetCustomerError;
/**
 * Error cuando hay problemas al obtener los clientes.
 */
class FlowGetCustomerListError extends FlowError {
    constructor(message) {
        super(`Error al obtener los clientes: ${message}`);
        this.name = 'FlowGetCustomersError';
    }
}
exports.FlowGetCustomerListError = FlowGetCustomerListError;
/**
 * Error cuando hay problemas al tratar de registrar la tarjeta de un cliente.
 */
class FlowRegisterCardError extends FlowError {
    constructor(message) {
        super(`Error al registrar la tarjeta: ${message}`);
        this.name = 'FlowRegisterCardError';
    }
}
exports.FlowRegisterCardError = FlowRegisterCardError;
/**
 * Error cuando hay problemas al tratar de ver el estado de la tarjeta de un cliente.
 */
class FlowRegisterCardStatusError extends FlowError {
    constructor(message) {
        super(`Error al ver el estado de la tarjeta: ${message}`);
        this.name = 'FlowRegisterCardStatusError';
    }
}
exports.FlowRegisterCardStatusError = FlowRegisterCardStatusError;
/**
 * Error cuando hay problemas al tratar de eliminar la tarjeta de un cliente.
 */
class FlowDeleteCardError extends FlowError {
    constructor(message) {
        super(`Error al eliminar la tarjeta: ${message}`);
        this.name = 'FlowDeleteCardError';
    }
}
exports.FlowDeleteCardError = FlowDeleteCardError;
/**
 * Error cuando hay problemas al tratar de obtener la tarjeta de un cliente.
 */
class FlowChargeCardError extends FlowError {
    constructor(message) {
        super(`Error al cargar la tarjeta: ${message}`);
        this.name = 'FlowChargeCardError';
    }
}
exports.FlowChargeCardError = FlowChargeCardError;
/**
 * Error cuando hay problemas al tratar de enviar cobro a un cliente.
 */
class FlowSendChargeCardError extends FlowError {
    constructor(message) {
        super(`Error al enviar cobro a la tarjeta: ${message}`);
        this.name = 'FlowSendChargeCardError';
    }
}
exports.FlowSendChargeCardError = FlowSendChargeCardError;
/**
 * Error cuando hay problemas al tratar de enviar varios cobros a varios clientes.
 */
class FlowSendMassiveChargeCardError extends FlowError {
    constructor(message) {
        super(`Error al enviar cobros masivos a tarjetas: ${message}`);
        this.name = 'FlowSendMassiveChargeCardError';
    }
}
exports.FlowSendMassiveChargeCardError = FlowSendMassiveChargeCardError;
/**
 * Error cuando hay problemas al tratar de ver el estado de varios cobros a varios
 */
class FlowMassiveChargeCardStatusError extends FlowError {
    constructor(message) {
        super(`Error al ver el estado de cobros masivos a tarjetas: ${message}`);
        this.name = 'FlowMassiveChargeCardStatusError';
    }
}
exports.FlowMassiveChargeCardStatusError = FlowMassiveChargeCardStatusError;
/**
 *  Error cuando hay problemas al tratar de hacer reversa en un cargo efectuado en la tarjeta de un cliente
 */
class FlowReverseChargeCardError extends FlowError {
    constructor(message) {
        super(`Error al hacer reversa de un cargo en la tarjeta: ${message}`);
        this.name = 'FlowReverseChargeCardError';
    }
}
exports.FlowReverseChargeCardError = FlowReverseChargeCardError;
/**
 * Error cuando hay problemas al listar paginadamente los cargos efectuados a un cliente
 */
class FlowListChargesCardError extends FlowError {
    constructor(message) {
        super(`Error al listar cargos efectuados en la tarjeta: ${message}`);
        this.name = 'FlowListChargesCardError';
    }
}
exports.FlowListChargesCardError = FlowListChargesCardError;
/**
 * Error cuando hay problemas al listar paginadamente los cargos fallidos a un cliente
 */
class FlowListFailedChargesCardError extends FlowError {
    constructor(message) {
        super(`Error al listar cargos fallidos en la tarjeta: ${message}`);
        this.name = 'FlowListFailedChargesCardError';
    }
}
exports.FlowListFailedChargesCardError = FlowListFailedChargesCardError;
/**
 * Error cuando hay problemas al listar paginadamente los cobros efectuados a un cliente
 */
class FlowListPaginatedSubscriptionsError extends FlowError {
    constructor(message) {
        super(`Error al listar suscripciones paginadas: ${message}`);
        this.name = 'FlowListPaginatedSubscriptionsError';
    }
}
exports.FlowListPaginatedSubscriptionsError = FlowListPaginatedSubscriptionsError;
/**
 *  Error cuando hay problemas al tratar de crear una suscripción
 */
class FlowCreatePlanError extends FlowError {
    constructor(message) {
        super(`Error al crear el plan: ${message}`);
        this.name = 'FlowCreatePlanError';
    }
}
exports.FlowCreatePlanError = FlowCreatePlanError;
/**
 * Error cuando hay problemas al tratar de obtener una suscripción
 */
class FlowGetPlanError extends FlowError {
    constructor(message) {
        super(`Error al obtener el plan: ${message}`);
        this.name = 'FlowGetPlanError';
    }
}
exports.FlowGetPlanError = FlowGetPlanError;
/**
 * Error cuando hay problemas al tratar de editar una suscripción
 */
class FlowEditPlanError extends FlowError {
    constructor(message) {
        super(`Error al editar el plan: ${message}`);
        this.name = 'FlowEditPlanError';
    }
}
exports.FlowEditPlanError = FlowEditPlanError;
/**
 * Error cuando hay problemas al tratar de eliminar una suscripción
 */
class FlowListPlansError extends FlowError {
    constructor(message) {
        super(`Error al listar los planes: ${message}`);
        this.name = 'FlowListPlansError';
    }
}
exports.FlowListPlansError = FlowListPlansError;
/**
 * Error cuando hay problemas al tratar de obtener una suscripción
 */
class FlowCreateSubscriptionToPlanError extends FlowError {
    constructor(message) {
        super(`Error al crear la suscripción al plan: ${message}`);
        this.name = 'FlowCreateSubscriptionToPlanError';
    }
}
exports.FlowCreateSubscriptionToPlanError = FlowCreateSubscriptionToPlanError;
/**
 * Error cuando hay problemas al tratar de obtener una subscripción por ID
 *
 */
class FlowGetSubscriptionBySubscriptionIdError extends FlowError {
    constructor(message) {
        super(`Error al obtener la suscripción por ID: ${message}`);
        this.name = 'FlowGetSubscriptionBySubscriptionIdError';
    }
}
exports.FlowGetSubscriptionBySubscriptionIdError = FlowGetSubscriptionBySubscriptionIdError;
/**
 * Error cuando hay problemas al tratar de obtener las subscripciones de un plan
 */
class FlowGetPlanSubscriptionsError extends FlowError {
    constructor(message) {
        super(`Error al obtener las suscripciones del plan: ${message}`);
        this.name = 'FlowGetPlanSubscriptionsError';
    }
}
exports.FlowGetPlanSubscriptionsError = FlowGetPlanSubscriptionsError;
/**
 * Error cuando hay problemas al tratar de obtener las subscripciones de un cliente
 */
class FlowUpdateSubscriptionTrialDaysError extends FlowError {
    constructor(message) {
        super(`Error al actualizar los días de prueba de la suscripción: ${message}`);
        this.name = 'FlowUpdateSubscriptionTrialDaysError';
    }
}
exports.FlowUpdateSubscriptionTrialDaysError = FlowUpdateSubscriptionTrialDaysError;
/**
 * Error cuando hay problemas al tratar de cancelar una suscripción
 */
class FlowCancelSubscriptionError extends FlowError {
    constructor(message) {
        super(`Error al cancelar la suscripción: ${message}`);
        this.name = 'FlowCancelSubscriptionError';
    }
}
exports.FlowCancelSubscriptionError = FlowCancelSubscriptionError;
class FlowAddDiscountToSubscriptionError extends FlowError {
    constructor(message) {
        super(`Error al agregar descuento a la suscripción: ${message}`);
        this.name = 'FlowAddDiscountToSubscription';
    }
}
exports.FlowAddDiscountToSubscriptionError = FlowAddDiscountToSubscriptionError;
class FlowRemoveDiscountFromSubscriptionError extends FlowError {
    constructor(message) {
        super(`Error al remover descuento de la suscripción: ${message}`);
        this.name = 'FlowRemoveDiscountFromSubscription';
    }
}
exports.FlowRemoveDiscountFromSubscriptionError = FlowRemoveDiscountFromSubscriptionError;
class FlowAddItemToSubscriptionError extends FlowError {
    constructor(message) {
        super(`Error al agregar item a la suscripción: ${message}`);
        this.name = 'FlowAddItemToSubscription';
    }
}
exports.FlowAddItemToSubscriptionError = FlowAddItemToSubscriptionError;
class FlowRemoveItemFromSubscriptionError extends FlowError {
    constructor(message) {
        super(`Error al remover item de la suscripción: ${message}`);
        this.name = 'FlowRemoveItemFromSubscription';
    }
}
exports.FlowRemoveItemFromSubscriptionError = FlowRemoveItemFromSubscriptionError;
class FlowChangeAssociatedPlanToSubscriptionError extends FlowError {
    constructor(message) {
        super(`Error al cambiar plan asociado a la suscripción: ${message}`);
        this.name = 'FlowChangeAssociatedPlanToSubscription';
    }
}
exports.FlowChangeAssociatedPlanToSubscriptionError = FlowChangeAssociatedPlanToSubscriptionError;
class FlowPreviewSubscriptionPlanChangeError extends FlowError {
    constructor(message) {
        super(`Error al previsualizar cambio de plan de la suscripción: ${message}`);
        this.name = 'FlowPreviewSubscriptionPlanChange';
    }
}
exports.FlowPreviewSubscriptionPlanChangeError = FlowPreviewSubscriptionPlanChangeError;
class FlowCancelScheduledPlanChangeError extends FlowError {
    constructor(message) {
        super(`Error al cancelar cambio plan programado: ${message}`);
        this.name = 'FlowCancelScheduledPlanChange';
    }
}
exports.FlowCancelScheduledPlanChangeError = FlowCancelScheduledPlanChangeError;
class FlowCreateAdditionalSubscriptionItemError extends FlowError {
    constructor(message) {
        super(`Error al crear item adicional de suscripción: ${message}`);
        this.name = 'FlowCreateAdditionalSubscriptionItemError';
    }
}
exports.FlowCreateAdditionalSubscriptionItemError = FlowCreateAdditionalSubscriptionItemError;
class FlowGetAdditionalSubscriptionItemError extends FlowError {
    constructor(message) {
        super(`Error al obtener item adicional de suscripción: ${message}`);
        this.name = 'FlowGetAdditionalSubscriptionItemError';
    }
}
exports.FlowGetAdditionalSubscriptionItemError = FlowGetAdditionalSubscriptionItemError;
class FlowEditAdditionalSubscriptionItemError extends FlowError {
    constructor(message) {
        super(`Error al editar item adicional de suscripción: ${message}`);
        this.name = 'FlowEditAdditionalSubscriptionItemError';
    }
}
exports.FlowEditAdditionalSubscriptionItemError = FlowEditAdditionalSubscriptionItemError;
class FlowDeleteAdditionalSubscriptionItemError extends FlowError {
    constructor(message) {
        super(`Error al eliminar item adicional de suscripción: ${message}`);
        this.name = 'FlowDeleteAdditionalSubscriptionItem';
    }
}
exports.FlowDeleteAdditionalSubscriptionItemError = FlowDeleteAdditionalSubscriptionItemError;
class FlowListAdditionalSubscriptionItemError extends FlowError {
    constructor(message) {
        super(`Error al listar items adicionales de suscripción: ${message}`);
        this.name = 'FlowListAdditionalSubscriptionItemError';
    }
}
exports.FlowListAdditionalSubscriptionItemError = FlowListAdditionalSubscriptionItemError;
class FlowCreateDiscountCouponError extends FlowError {
    constructor(message) {
        super(`Error al crear cupón de descuento: ${message}`);
        this.name = 'FlowCreateDiscountCouponError';
    }
}
exports.FlowCreateDiscountCouponError = FlowCreateDiscountCouponError;
class FlowEditDiscountCouponError extends FlowError {
    constructor(message) {
        super(`Error al editar cupón de descuento: ${message}`);
        this.name = 'FlowEditDiscountCouponError';
    }
}
exports.FlowEditDiscountCouponError = FlowEditDiscountCouponError;
class FlowDeleteDiscountCouponError extends FlowError {
    constructor(message) {
        super(`Error al eliminar cupón de descuento: ${message}`);
        this.name = 'FlowDeleteDiscountCouponError';
    }
}
exports.FlowDeleteDiscountCouponError = FlowDeleteDiscountCouponError;
class FlowGetDiscountCouponError extends FlowError {
    constructor(message) {
        super(`Error al obtener cupón de descuento: ${message}`);
        this.name = 'FlowGetDiscountCouponError';
    }
}
exports.FlowGetDiscountCouponError = FlowGetDiscountCouponError;
class FlowListDiscountCouponsError extends FlowError {
    constructor(message) {
        super(`Error al listar cupones de descuento: ${message}`);
        this.name = 'FlowListDiscountCouponsError';
    }
}
exports.FlowListDiscountCouponsError = FlowListDiscountCouponsError;
class FlowGetInvoiceDataError extends FlowError {
    constructor(message) {
        super(`Error al obtener datos de la factura: ${message}`);
        this.name = 'FlowGetInvoiceDataError';
    }
}
exports.FlowGetInvoiceDataError = FlowGetInvoiceDataError;
class FlowCancelInvoicePendingPaymentError extends FlowError {
    constructor(message) {
        super(`Error al cancelar factura pendiente de pago: ${message}`);
        this.name = 'FlowCancelInvoicePendingPaymentError';
    }
}
exports.FlowCancelInvoicePendingPaymentError = FlowCancelInvoicePendingPaymentError;
class FlowRecordExternalPaymentAndMarkInvoicePaidError extends FlowError {
    constructor(message) {
        super(`Error al registrar pago externo y marcar factura pagada: ${message}`);
        this.name = 'FlowRecordExternalPaymentAndMarkInvoicePaidError';
    }
}
exports.FlowRecordExternalPaymentAndMarkInvoicePaidError = FlowRecordExternalPaymentAndMarkInvoicePaidError;
class FlowGetOverdueInvoicesError extends FlowError {
    constructor(message) {
        super(`Error al obtener facturas vencidas: ${message}`);
        this.name = 'FlowGetOverdueInvoicesError';
    }
}
exports.FlowGetOverdueInvoicesError = FlowGetOverdueInvoicesError;
class FlowRetryOverdueInvoicePaymentError extends FlowError {
    constructor(message) {
        super(`Error al reintentar pago de factura vencida: ${message}`);
        this.name = 'FlowRetryOverdueInvoicePaymentError';
    }
}
exports.FlowRetryOverdueInvoicePaymentError = FlowRetryOverdueInvoicePaymentError;
class FlowGetLiquidationsByDateRangeError extends FlowError {
    constructor(message) {
        super(`Error al obtener liquidaciones por rango de fecha: ${message}`);
        this.name = 'FlowGetLiquidationsByDateRangeError';
    }
}
exports.FlowGetLiquidationsByDateRangeError = FlowGetLiquidationsByDateRangeError;
class FlowCreateAssociatedMerchantError extends FlowError {
    constructor(message) {
        super(`Error al crear comercio asociado: ${message}`);
        this.name = 'FlowCreateAssociatedMerchantError';
    }
}
exports.FlowCreateAssociatedMerchantError = FlowCreateAssociatedMerchantError;
class FlowDeleteAssociatedMerchantError extends FlowError {
    constructor(message) {
        super(`Error al eliminar comercio asociado: ${message}`);
        this.name = 'FlowDeleteAssociatedMerchantError';
    }
}
exports.FlowDeleteAssociatedMerchantError = FlowDeleteAssociatedMerchantError;
class FlowGetAssociatedMerchantError extends FlowError {
    constructor(message) {
        super(`Error al obtener comercio asociado: ${message}`);
        this.name = 'FlowGetAssociatedMerchantError';
    }
}
exports.FlowGetAssociatedMerchantError = FlowGetAssociatedMerchantError;
